### PR TITLE
LSTM GRIN - code

### PR DESCRIPTION
**Pull request by wzoustanford**
This PR includes work to implement Global Recurrent Inhibition Networks in a recurrent fashion. 

We implement a replay buffer that stores and samples sequences of {(s, a, n_s, r)}. 
The GRIN network takes activations from previous time-steps and produces masks in the ensuing step. 
The recurrent memory is handled by an LSTM. 
The implementation now only takes the combined output (OR) of the critic network, and masks the same location for the next time step. 

The implementation is in JAX and optimized for performance. The code runs at 0.15 steps (transitions) per second, making 1e8 steps to finish within 4 days. There is room to optimize the code and algorithm to run at a faster speed; for example, there is overhead with the LSTM structure.

### CODE REVIEW

File: LSTM_INTEGRATION_SUMMARY.md
Comment (rainx0r) [inline LSTM_INTEGRATION_SUMMARY.md:1]: Please do not commit Claude Code MD files

File: LSTM_TEST_SUMMARY.md
Comment (rainx0r) [inline LSTM_TEST_SUMMARY.md:1]: Please do not commit Claude Code MD files

File: metaworld_algorithms/config/nn.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 @dataclass(frozen=True, kw_only=True)
 class NeuralNetworkConfig:
```
Comment (rainx0r) [inline metaworld_algorithms/config/nn.py:8]: The defaults for NeuralNetworkConfig should not change.

File: metaworld_algorithms/config/optim.py [hunk 2]
```diff
@@ -24,7 +24,7 @@ def spawn(self) -> optax.GradientTransformation:
         # From https://github.com/araffin/sbx/blob/master/sbx/ppo/policies.py#L120
         optim_kwargs = {}
         if self.optimizer == Optimizer.Adam:
-            optim_kwargs["eps"] = 1e-5
+            optim_kwargs["eps"] = 1e-8
```
Comment (rainx0r) [inline metaworld_algorithms/config/optim.py:27]: This is also a needless change. The original code has a comment citing why it's set to this value. Can you provide a reference for why it should be lowered?

File: metaworld_algorithms/config/rl.py [hunk 3]
```diff
@@ -30,6 +30,7 @@ class OffPolicyTrainingConfig(TrainingConfig):
     warmstart_steps: int = int(4e3)
     buffer_size: int = int(1e6)
     batch_size: int = 1280
+    seq_len: int = 100
```
Comment (rainx0r) [inline metaworld_algorithms/config/rl.py:33]: This parameter is not needed in all OffPolicy algorithms and should not be here. You should subclass OffPolicyTrainingConfig to something like RecurrentOffPolicyTrainingConfig to add it

File: metaworld_algorithms/nn/moore.py [hunk 4]
```diff
@@ -3,15 +3,138 @@
 import jax
 import jax.numpy as jnp
 from jaxtyping import Array, Float
+from typing import Tuple
```
Comment (rainx0r) [inline metaworld_algorithms/nn/moore.py:6]: Use `tuple` to avoid this extra import

File: metaworld_algorithms/nn/moore.py [hunk 5]
```diff
@@ -3,15 +3,138 @@
...
 from metaworld_algorithms.config.nn import MOOREConfig
 
 from .base import MLP
 
+# --- OPTIMIZED FUNCTION USING JAX.LAX.SCAN ---
 
 def orthogonal_1d(
     x: Float[Array, "batch_size num_experts dim"], num_experts: int
 ) -> Float[Array, "batch_size num_experts dim"]:
+
```
Comment (rainx0r) [inline metaworld_algorithms/nn/moore.py:17]: Why did this function need to be rewritten? Can you provide justification in the form of a script comparing the performance (and speed) between this and my original implementation?

File: metaworld_algorithms/nn/moore.py [hunk 6]
```diff
@@ -67,12 +205,32 @@ def __call__(self, x: jax.Array) -> jax.Array:
...
         )(x)
+        # Orthogonalize expert outputs for better representation diversity
         experts_out = orthogonal_1d(experts_out, num_experts=self.config.num_experts)
+        # Combine experts using task embedding weights
         features_out = jnp.einsum("bnk,bn->bk", experts_out, task_embedding)
         features_out = jax.nn.tanh(features_out)
-        self.sow("intermediates", "torso_output", features_out)
 
-        # MH
+        # LSTM Integration: Apply mask if provided
```
Comment (rainx0r) [inline metaworld_algorithms/nn/moore.py:214]: The `self.sow()` call is actually needed here and was an intentional choice in the original code, I'm not sure why 1) it was first attempted to be moved to be conditional on the `store` flag 2) removed entirely apparently.

File: metaworld_algorithms/rl/algorithms/base.py [hunk 7]
```diff
@@ -632,11 +633,21 @@ def train(
                         },
                         step=total_steps,
                     )
-
+            
             if global_step > config.warmstart_steps:
                 # Update the agent with data
-                data = replay_buffer.sample(config.batch_size)
+                print(f"global_step:{global_step};  total_step:{total_steps}")
```
Comment (rainx0r) [inline metaworld_algorithms/rl/algorithms/base.py:639]: I'm not sure these extra timings are needed especially for batch sampling time.

File: metaworld_algorithms/rl/algorithms/base.py [hunk 8]
```diff
@@ -588,7 +588,8 @@ def train(
 
         for global_step in range(start_step, config.total_steps // envs.num_envs):
             total_steps = global_step * envs.num_envs
-
+            if global_step > config.warmstart_steps:
```
Comment (rainx0r) [inline metaworld_algorithms/rl/algorithms/base.py:591]: I don't think this is needed, at least not in the general case. I'd suggest implementing a new `RecurrentOffPolicyAlgorithm` class that has extra logic for handling the seq_len and adding it to total steps, but also it should probably just be added to the line above (590)?

File: metaworld_algorithms/rl/algorithms/mtsac.py
Comment (rainx0r) [inline metaworld_algorithms/rl/algorithms/mtsac.py:1]: Just implement this as a new algorithm file, don't pollute mtsac.py. But also the code quality seems very dubious compared to the rest of the codebase and I'd be hesitant to introduce such an implementation into the repository. Why is the LSTM forward pass and logic defined at the algorithm loss function level and not at the network level? The algorithm should be agnostic as to the network's structure. The only thing you really need to add to MTSAC to make it work with recurrent networks is the ability for it to also sample network hidden states from the replay buffer I think.

File: metaworld_algorithms/rl/algorithms/mtsac_before_gemini.py
Comment (rainx0r) [inline metaworld_algorithms/rl/algorithms/mtsac_before_gemini.py:1]: Duplicate implementations of the same algorithm just because the newer implementation by the LLM is bad are not acceptable. The updated implementation should be good enough to replace the original one, which currently it is not so I understand the need for this file, but generally this should not happen. Once again I think you should probably not touch `mtsac.py` and instead make a new file that's maybe `recurrent_mtsac.py` or something and has the minimal changes necessary to add recurrent network logic into MTSAC. Then all the recurrent network forward passes and masking etc can be implemented at the network forward pass level (maybe make a new file that's `nn/moore_recurrent.py` and add the appropriate network configs for it etc).

File: metaworld_algorithms/rl/networks.py [hunk 9]
```diff
@@ -252,6 +252,48 @@ def __call__(self, state: jax.Array) -> jax.Array:
                 "Value prediction as classification is not supported yet."
             )
 
+"""
```
Comment (rainx0r) [inline metaworld_algorithms/rl/networks.py:255]: What is the purpose of this comment?

File: metaworld_algorithms/rl/networks.py [hunk 10]
```diff
@@ -214,7 +214,7 @@ class QValueFunction(nn.Module):
     config: QValueFunctionConfig
 
     @nn.compact
-    def __call__(self, state: jax.Array, action: jax.Array) -> jax.Array:
+    def __call__(self, state: jax.Array, action: jax.Array, mask: jax.Array | None = None, store: bool = False) -> jax.Array:
```
Comment (rainx0r) [inline metaworld_algorithms/rl/networks.py:217]: The RL network interfaces should be generally agnostic to lower level implementation details. The only thing that should be necessary here is taking in a network hidden state or something to make it work with recurrent architectures.

File: test_lstm.py
Comment (rainx0r) [inline test_lstm.py:1]: If you're going to add tests, do so in a `tests/` directory preferably with a subfolder structure mirroring the project to reflect the granularity of what's being tested. This should go into `tests/nn/test_lstm.py` or something. Also instead of running through main() you should use Pytest as the test runner.

File: test_lstm_training.py
Comment (rainx0r) [inline test_lstm_training.py:1]: If you're going to add tests, do so in a `tests/` directory preferably with a subfolder structure mirroring the project to reflect the granularity of what's being tested. This should go into `tests/nn/test_lstm_training.py` or something. Also instead of running through main() you should use Pytest as the test runner.

**Review by rainx0r — CHANGES_REQUESTED** [review]
Thank you for your pull request. I think the code quality is insufficient to be merged and needs a lot of changes that I'm unsure if your AI code generation stack is capable of handling adequately, but I thought I'd humour it. Out of curiosity, what tools are you using? Gemini is mentioned so is this Gemini 3 Pro or some other model? (the .MD files sort of imply Claude Code)

My goal with this repository is to keep it as minimal and composable as possible for didactic and practical purposes. While having read through the PR I'm sure the code is functional (probably? some of the MOORE changes look weird), it does not meet that criteria.

I think what's needed as a first step is to add general "recurrent-network" functionality to off policy and on policy algorithms (for on policy inspiration can be taken from the RL2 implementation in the repository, for off policy it should be as simple as adding hidden states into the buffer and making it so the buffer can take in arbitrary batch dimensions, and adding in a new recurrent training loop to store sequences into the buffer as you've done). Then you can implement GRIN as simply another network architecture that plugs into the new general recurrent network functionality.

**Comment by wzoustanford** [comment]
Hi rainx0r, I'm sorry, this is a mistake on my end. The PR should have been made on the forked version of the code. 

Thank you for making this code available for research. 

While I like it the way you would humor the code changes, I would love to collaborate on this repo in December or January. I'd have to delete this PR for now, but let me make a new one later in December. Thank you. 

We're working on research.

**Comment by wzoustanford** [comment]
@rainx0r I'd appreciate it if you'd keep discussions in this PR private and confidential before mid-December. I will reach out then. Thank you!
